source/azure_iot.c

```diff
@@ -0,0 +1,412 @@
+    {
+        /* Some padding is needed.  */
+
+        /* Calculate the number of pad characters.  */
+        ulPad = ( ulLength * 6 ) % 24;
+        ulPad = ( 24 - ulPad ) / 6;
+        ulPad = ulPad - 1;
+
+        /* Adjust the ulLength to pickup the character fraction.  */
+        ulLength++;
+    }
+
+    if( base64name_size <= ulLength )
+    {
+        return eAzureIoTOutOfMemory;
+    }
+
+    /* Setup index into the pcBase64name.  */
+    j = 0;
+
+    /* Compute the pcBase64name.  */
+    ulStep = 0;
+    i = 0;
+
+    while( j < ulLength )
+    {
+        /* Determine which step we are in.  */
+        if( ulStep == 0 )
+        {
+            /* Use first 6 bits of name character for index.  */
+            pcBase64name[ j++ ] = ( char ) _cAzureIoTBase64Array[ ( ( uint8_t ) pucName[ i ] ) >> 2 ];
+            ulStep++;
+        }
+        else if( ulStep == 1 )
+        {
+            /* Use last 2 bits of name character and first 4 bits of next name character for index.  */
+            pcBase64name[ j++ ] = ( char ) _cAzureIoTBase64Array[ ( ( ( ( uint8_t ) pucName[ i ] ) & 0x3 ) << 4 ) |
+                                                                  ( ( ( uint8_t ) pucName[ i + 1 ] ) >> 4 ) ];
+            i++;
+            ulStep++;
+        }
+        else if( ulStep == 2 )
+        {
+            /* Use last 4 bits of name character and first 2 bits of next name character for index.  */
+            pcBase64name[ j++ ] = ( char ) _cAzureIoTBase64Array[ ( ( ( ( uint8_t ) pucName[ i ] ) & 0xF ) << 2 ) |
+                                                                  ( ( ( uint8_t ) pucName[ i + 1 ] ) >> 6 ) ];
+            i++;
+            ulStep++;
+        }
+        else /* Step 3 */
+        {
+            /* Use last 6 bits of name character for index.  */
+            pcBase64name[ j++ ] = ( char ) _cAzureIoTBase64Array[ ( ( ( uint8_t ) pucName[ i ] ) & 0x3F ) ];
+            i++;
+            ulStep = 0;
+        }
+    }
+
+    /* Determine if the index needs to be advanced.  */
+    if( ulStep != 3 )
+    {
+        i++;
+    }
+
+    /* Now add the PAD characters.  */
+    while( ( ulPad-- ) && ( j < base64name_size ) )
+    {
+        /* Pad pcBase64name with '=' characters.  */
+        pcBase64name[ j++ ] = '=';
+    }
+
+    /* Put a NULL character in.  */
+    pcBase64name[ j ] = 0;
+
+    return eAzureIoTSuccess;
+}
+/*-----------------------------------------------------------*/
+
+AzureIoTResult_t AzureIoT_Init()
+{
+    #if ( LIBRARY_LOG_LEVEL >= LOG_INFO )
+        az_log_set_message_callback( prvAzureIoTLogListener );
+    #endif
+
+    return eAzureIoTSuccess;
+}
+/*-----------------------------------------------------------*/
+
+void AzureIoT_Deinit()
+{
+    AZLogDebug( ( "AzureIoT_Deinit called \r\n" ) );
+}
+/*-----------------------------------------------------------*/
+
+AzureIoTResult_t AzureIoT_MessagePropertiesInit( AzureIoTMessageProperties_t * pxMessageProperties,
```

Comment:
good point. we should probably discuss here. @hihigupt you have any opinion? I don't see any reason why we can't just make it a pointer only (no handle) since we don't do any malloc'ing?

Reply:
Making an extra type for pointer to control block is fine with me.

Reply:
discussed offline. Removing `_Handle` types since user allocates the structs and hands them to us. Giving us pointers to structs makes sense in that case. If we were `malloc`'ing and returning to the user it would probably make more sense to use handles.

Reply:
https://github.com/Azure/azure-iot-middleware-freertos/pull/67
PR to tackle